### PR TITLE
Download latest version if user uses regex and TFENV_AUTO_INSTALL is true

### PR DESCRIPTION
**Pull request by aston-r**
Closes: https://github.com/tfutils/tfenv/issues/216

flow:
- set latest:^0.12 in .terraform-version
- check latest available remove version
- compare the latest available remote version and the latest installed version
- install latest available remote version if it is newer

### CODE REVIEW

**Comment by aston-r** [comment]
@kamatama41  @anthraxx  Could you please review the changes or assign reviewers?

**Comment by aston-r** [comment]
Could you please review the changes? @kamatama41 @iamhsa  @Zordrak @dooreelko

File: libexec/tfenv-version-name [hunk 1]
```diff
@@ -86,7 +86,7 @@ if [[ "${TFENV_VERSION}" =~ ^latest.*$ ]]; then
     regex="${TFENV_VERSION##*\:}";
     log 'debug' "'latest' keyword uses regex: ${regex}";
   else
-    regex='.*'; # Just saves a seperate command below without the grep
+    regex='.*'; # Just saves a separate command below without the grep
```
Comment (dooreelko) [inline libexec/tfenv-version-name:89]: nice

File: libexec/tfenv-version-name [hunk 2]
```diff
@@ -96,13 +96,23 @@ if [[ "${TFENV_VERSION}" =~ ^latest.*$ ]]; then
...
+          TFENV_VERSION="${version}"
+        fi
+    else
+      log 'error' "No versions matching '${requested}' found in remote";
+    fi
   else
-    log 'error' "No installed versions of terraform matched '${TFENV_VERSION}'";
-  fi;
-
+    [[ -n "${version}" ]] && TFENV_VERSION="${version}" \
```
Comment (dooreelko) [inline libexec/tfenv-version-name:113]: this used to be a cleaner if/else and i think this codebase prefers explicit if/else for actions but allows `condition && log 'info' 'ok' || log 'error' 'not ok' pattern.

File: libexec/tfenv-version-name [hunk 3]
```diff
@@ -96,13 +96,23 @@ if [[ "${TFENV_VERSION}" =~ ^latest.*$ ]]; then
     | grep -e "${regex}" \
     | head -n 1)";
 
-  if [ -n "${version}" ]; then
-    log 'debug' "Version selected: ${version}";
-    TFENV_VERSION="${version}"
+  if [[ "${TFENV_AUTO_INSTALL:-true}" == "true" ]]; then
```
[the remark this answers is not in the document]
Comment (dooreelko) [reply libexec/tfenv-version-name:99]: ah, apparently the default is true, so disregard

**Review by dooreelko — APPROVED** [review]
aww, i have the right to review?
lgtm, i'd fixed that if/else at the end to be more legible

**Review by dooreelko — APPROVED** [review]

**Comment by dooreelko** [comment]
looks like my review doesn't qualify, still :shrug:

**Comment by aston-r** [comment]
@Zordrak  Could you please review the changes?

**Review by Zordrak — DISMISSED** [review]

**Review by Zordrak — APPROVED** [review]

**Review by Zordrak — DISMISSED** [review]

**Comment by Zordrak** [comment]
I've done my best to resolve conflicts from other pushes, but tests are still failing

**Review by Zordrak — APPROVED** [review]

**Comment by Zordrak** [comment]
You exposed an existing bug - which I've now fixed.
Now the existing tests pass I will merge this - although I would really appreciate it if you could offer up some tests for the auto-install logic, as I've not the time right now.

**Pull request merged**